### ISSUE TITLE
performance isuue

### ISSUE THREAD

**Issue by tonghe90** [HuguesTHOMAS/KPConv#4]
Hi @HuguesTHOMAS :
    Thanks for sharing the code, in order to get the performance reported in the paper, (iou s3dis area5), I change the self.validation_split = 4 in S3DIS.py. But I cann't get the performance reported in the paper (miou 66). Do you have any suggestion? I haven't finished training, but the miou seems not going up.

Step 00089533 L_out=0.049 L_reg=0.033 L_p=0.000 Acc=0.98 ---  709.37 ms/batch (Averaged) 
Step 00089535 L_out=0.061 L_reg=0.033 L_p=0.000 Acc=0.98 ---  711.77 ms/batch (Averaged)
Step 00089537 L_out=0.043 L_reg=0.033 L_p=0.000 Acc=0.98 ---  717.27 ms/batch (Averaged)
Step 00089539 L_out=0.058 L_reg=0.033 L_p=0.000 Acc=0.98 ---  721.15 ms/batch (Averaged)
Validation : 0.0% (timings : 165.89 6.77)
Validation : 6.0% (timings : 276.87 20.58)
Validation : 10.0% (timings : 301.69 28.65)
Validation : 16.0% (timings : 306.19 33.44)
Validation : 22.0% (timings : 313.93 38.35)
Validation : 26.0% (timings : 333.55 41.70)
Validation : 30.0% (timings : 350.38 45.12)
Validation : 32.0% (timings : 383.41 46.56)
Validation : 38.0% (timings : 379.15 50.89)
Validation : 42.0% (timings : 389.80 52.16)
Validation : 46.0% (timings : 404.30 56.77)
Validation : 48.0% (timings : 447.05 57.91)
Validation : 54.0% (timings : 432.56 60.22)
Validation : 60.0% (timings : 424.93 60.28)
Validation : 66.0% (timings : 414.28 59.36)
Validation : 72.0% (timings : 399.32 57.45)
S3DIS mean IoU = 56.9%

**Comment by HuguesTHOMAS** [comment]
Hi @tonghe90 ,

Actually, the value reported in the console does not correpond to the real validation score. This value (56.9%) is the "network output validation score". 

If you report to the paper (Section 4.2, "Pipeline for real scene segmentation"), you see that the network is used on small spheres randomly chosen across the scene. Therefore, the validation score can be measured in two ways:

- "network output validation score": stack the predictions of the spheres chosen for this validation step, and computing confusion of this subset of points
- "dataset validation score": reproject the predictions of each sphere on the dataset (using a probability voting scheme), and computing confusion on the whole dataset

As you can see the first measure is not the real score, as it is not computed on the dataset points. It is an average score of the network output. However it would be too long too compute the whole dataset confusion at every epoch, so we use this network output value to see the convergence.

To see the real "dataset validation score", you can look at the log folder. You will find the whole dataset predictions at different epochs in folders called val_preds_XXX. The most convenient way to visualize the scores is to use the `plot_convergence.py` script. The network output validation is plotted as dashed line, and the real score as solid line.

**Comment by kentangSJTU** [comment]
Hi, @HuguesTHOMAS 
    I tried to use your test code on S3DIS Area 5 after 120k training steps (corresponding to around 300 epochs). I noticed that the "network output validation score" is similar to "dataset validation score". Both values are around 60%. 

    From my understanding, the first value is calculated only with points in the KDTree (i.e. subsampled using grid sampling with a grid size of 0.04 meters) and the second value is calculated by projecting all the points within one grid to the same label. Is that understanding correct? And should the results be very similar?

    Thanks a lot for your reply~

Best,
Ken

**Comment by HuguesTHOMAS** [comment]
@kentangSJTU,

Their is a difference between what happens in the validation and what happens in the test. 

* In validation we have the score computed directly on the input spheres, and the score computed on the reprojections (which correspond to the _Confusion on full clouds_ in the test script). These two scores are described in my last message.

* In test we do not compute the score on the input spheres (it is quite useless during test, as it was only usefull to see the convergence during validation). We only keep track of the reprojected score, and your understanding is correct. The _Confusion on sub clouds_ is computed on the points of the KDTree and the _Confusion on full clouds_ is interpolated (nearest interpolation) to the full clouds.

For your second question, it is expected that the two test scores are similar, because the density of points in S3DIS is quite homogeneous, so the scores computed on the subsampled cloud are close to the scores computed on the whole cloud. Only the scores on the input spheres in the validation are expected to be very different.

In the end the only score that matters is the _Confusion on full clouds_. Whether you look at the one from the validation or from the test, it is the only one that can be compared to other methods.

**Comment by kentangSJTU** [comment]
> @kentangSJTU,
> 
> Their is a difference between what happens in the validation and what happens in the test.
> 
> * In validation we have the score computed directly on the input spheres, and the score computed on the reprojections (which correspond to the _Confusion on full clouds_ in the test script). These two scores are described in my last message.
> * In test we do not compute the score on the input spheres (it is quite useless during test, as it was only usefull to see the convergence during validation). We only keep track of the reprojected score, and your understanding is correct. The _Confusion on sub clouds_ is computed on the points of the KDTree and the _Confusion on full clouds_ is interpolated (nearest interpolation) to the full clouds.
> 
> For your second question, it is expected that the two test scores are similar, because the density of points in S3DIS is quite homogeneous, so the scores computed on the subsampled cloud are close to the scores computed on the whole cloud. Only the scores on the input spheres in the validation are expected to be very different.
> 
> In the end the only score that matters is the _Confusion on full clouds_. Whether you look at the one from the validation or from the test, it is the only one that can be compared to other methods.

Hi @HuguesTHOMAS,

Thanks a lot for your clear answers. I think I am clear about my questions. I will try to get the results after all training steps and compare with the reported results in your paper. BTW, is the provided architecture for S3DIS the one used in your experiments to produce the best results? Thanks for reply~

Best,
Ken

**Comment by HuguesTHOMAS** [comment]
Yes, the architecture and the parameters defined in `training_S3DIS` are the one I used for my experiments with rigid kernels. If you want deformable kernels, you just have to change the name of the layers in the architecture list (use "resnetb_deformable" and "resnetb_deformable_strided").

You can use the script called `plot_convergence` to show the validation scores, it can be run during training to see how things are going. You should see the score reprojected on the full clouds as a solid line and the score on validation spheres as a dashed line. The first one is only computed once every 50 epochs (so you won't see it at the beginning of the training), and the second one at each epoch.

**Comment by kentangSJTU** [comment]
> Yes, the architecture and the parameters defined in `training_S3DIS` are the one I used for my experiments with rigid kernels. If you want deformable kernels, you just have to change the name of the layers in the architecture list (use "resnetb_deformable" and "resnetb_deformable_strided").
> 
> You can use the script called `plot_convergence` to show the validation scores, it can be run during training to see how things are going. You should see the score reprojected on the full clouds as a solid line and the score on validation spheres as a dashed line. The first one is only computed once every 50 epochs (so you won't see it at the beginning of the training), and the second one at each epoch.

Thanks a lot for your reply~
